### PR TITLE
[SPARK-16649][SQL] Push partition predicates down into metastore for OptimizeMetadataOnlyQuery

### PR DESCRIPTION
**Pull request by lianhuiwang**
## What changes were proposed in this pull request?

SPARK-6910 has supported for pushing partition predicates down into the metastore for table scan. So it also should push partition predicates down into metastore for OptimizeMetadataOnlyQuery.
## How was this patch tested?

add unit tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala [hunk 1]
```diff
@@ -167,6 +168,21 @@ abstract class ExternalCatalog {
       table: String,
       partialSpec: Option[TablePartitionSpec] = None): Seq[CatalogTablePartition]
 
+  /**
+   * Returns partitions filtered by predicates for the given table, It just work for Hive.
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: ". It just works"

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala [hunk 2]
```diff
@@ -167,6 +168,21 @@ abstract class ExternalCatalog {
...
+   *
+   * The filters Expressions may optionally be provided to filter the partitions returned.
+   * For instance, if there exist partitions (a='1', b='2'), (a='1', b='3') and (a='2', b='4'),
+   * then the filters (a='1') will return the first two only.
+   * @param db database name
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalog.scala]: New line before `@param`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala [hunk 3]
```diff
@@ -632,6 +632,23 @@ class SessionCatalog(
   }
 
   /**
+   * Returns partitions filtered by predicates for the given table, It just work for Hive.
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Same as above

Comment (jaceklaskowski) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/SessionCatalog.scala]: Why do you copy the scaladoc?

**Comment by lianhuiwang** [comment]
cc @cloud-fan @rxin @hvanhovell

**Comment by cloud-fan** [comment]
seems we already have this?

**Comment by lianhuiwang** [comment]
@cloud-fan OK, I will close it. Thanks.

**Pull request closed without merging**